src/rate_limiter.cc

```diff
@@ -775,52 +757,89 @@ RateLimiter::ResourceManager::AddModelInstance(
+  // Find instance from model resources.
+  std::lock_guard<std::mutex> lk1(model_resources_mtx_);
   const auto& itr = model_resources_.find(instance);
   if (itr == model_resources_.end()) {
     return Status(
         Status::Code::INTERNAL, "Can not find the instance to remove");
   }
+  // Check if max resources need to be updated.
+  bool update_needed = false;
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  for (const auto& resource_device_map : itr->second) {
+    auto ditr = max_resources_.find(resource_device_map.first);
+    if (ditr != max_resources_.end()) {
+      for (const auto resource : resource_device_map.second) {
+        auto ritr = ditr->second.find(resource.first);
+        if (ritr != ditr->second.end() && ritr->second >= resource.second) {
+          update_needed = true;
+          break;
+        }
+      }
+    }
+    if (update_needed) {
+      break;
+    }
+  }
+  // Remove instance from model resources.
   model_resources_.erase(instance);
-  return Status::Success;
+  // Re-compute max resource if needed.
+  if (update_needed) {
+    ComputeResourceLimits();
+  }
+  return ParseAndValidateResources();
 }
 
-Status
-RateLimiter::ResourceManager::UpdateResourceLimits()
+void
+RateLimiter::ResourceManager::ComputeResourceLimits()
 {
-  std::lock_guard<std::mutex> lk1(model_resources_mtx_);
-  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  // Obtain the maximum resource across all the instances and use it as the
+  // default available.
   max_resources_.clear();
-  // Obtain the maximum resource across all the instances
-  // and use it as the default available.
   for (const auto& instance_resources : model_resources_) {
-    for (const auto& resource_device_map : instance_resources.second) {
-      auto ditr = max_resources_.find(resource_device_map.first);
-      if (ditr == max_resources_.end()) {
-        ditr =
-            max_resources_
-                .emplace(resource_device_map.first, resource_device_map.second)
-                .first;
-      } else {
-        for (const auto resource : resource_device_map.second) {
-          auto ritr = ditr->second.find(resource.first);
-          if (ritr == ditr->second.end()) {
-            ritr = ditr->second.emplace(resource.first, resource.second).first;
-          } else {
-            if (ritr->second < resource.second) {
-              ritr->second = resource.second;
-            }
+    UpdateMaxResource(instance_resources.second);
+  }
+}
+
+void
+RateLimiter::ResourceManager::UpdateMaxResource(
+    const ResourceMap& instance_resource_map)
+{
+  for (const auto& resource_device_map : instance_resource_map) {
+    auto ditr = max_resources_.find(resource_device_map.first);
+    if (ditr == max_resources_.end()) {
+      ditr = max_resources_
+                 .emplace(resource_device_map.first, resource_device_map.second)
+                 .first;
+    } else {
+      for (const auto resource : resource_device_map.second) {
```

Comment:
```suggestion
      for (const auto& resource : resource_device_map.second) {
```

Reply:
Updated: [Minor fix and return error on instance remove failure](https://github.com/triton-inference-server/core/pull/208/commits/33a5369309a117e216614f91ca96df7b95cadedb)